### ISSUE TITLE
Unusable with `react-modal` (no `form` dom element → no `submit` event bound).

### ISSUE THREAD

**Issue by futpib** [esamattis/react-zorm#15]
`react-modal` is usually always mounted, but it's children are not in the DOM until after it's `isOpen` prop turns `true`.

My guess is this should be fixable by changing `useEffect` to `useLayoutEffect` in `useZorm` here:
https://github.com/esamattis/react-zorm/blob/46c32bc2314e57bd0032d3e16053dee4149a9717/packages/react-zorm/src/use-zorm.tsx#L64
and recommending that users pass `setupListeners: isOpen` to `useZorm`.

Also this workaround works for me:

```typescript
// useZorm.ts
import { useLayoutEffect, useState } from "react";
import { Zorm, useZorm as useZormBase } from "react-zorm";
import { UseZormOptions } from "react-zorm/dist/use-zorm";
import { ZodType } from "zod";

export function useZorm<Schema extends ZodType<any>>(
	formName: string,
	schema: Schema,
	options?: UseZormOptions<ReturnType<Schema["parse"]>>,
): Zorm<Schema> {
	const [ setupListeners, setSetupListeners ] = useState(options?.setupListeners);

	useLayoutEffect(() => {
		setSetupListeners(options?.setupListeners);
	}, [
		options?.setupListeners,
	]);

	return useZormBase(formName, schema, {
		...options,
		setupListeners,
	});
}
```

```typescript
// MyModal.tsx
import { z } from 'zod';
import Modal from "react-modal";
import { useZorm } from './useZorm';

const formSchema = z.object({
	email: z.string().email(),
});

export interface MyModalProps {
	isOpen: boolean;
	onRequestClose: () => void;
}

export function MyModal({
	isOpen,
	onRequestClose,
}: MyModalProps) {
	const zo = useZorm('myModalForm', formSchema, {
		setupListeners: isOpen,
		onValidSubmit(event) {
			// never called, instead native form submit happens (page reload)
			debugger;
			event.preventDefault();
		},
	});

	return (
		<Modal
			isOpen={isOpen}
			onRequestClose={onRequestClose}
		>
			<form
				ref={zo.ref}
			>
				<input
					name={zo.fields.email()}
					type="email"
				/>

				<button
					type="submit"
				>
					submit
				</button>
			</form>
		</Modal>
	);
}
```

**Comment by esamattis** [comment]
Hey thanks for reporting!

I think this can be fixed by using ref callback and not ref objects as we are now currently using because ref callbacks get called even when the dom element is created at a later point of time.

I made a draft pr of it at #16 but is not ready for merging since it breaks `useValue()` and I think we need to do a breaking release to fix this since `useValue()` uses the ref object.

Anyway I made a dev release of it which you can install with:

```
npm install --save-exact react-zorm@0.4.0-dev.970757e413
```

Can you test it out and check if it fixes your issue?  Thanks.

**Comment by esamattis** [comment]
This should be now fixed in v0.5.0

Changelog and migration guide here https://github.com/esamattis/react-zorm/blob/master/packages/react-zorm/CHANGELOG.md#v050
